### PR TITLE
Baselines dqn waypoint ppo new reward pid vae refactor

### PR DESCRIPTION
**Pull request by hitesh11**
Merged baselines-dqn-waypoint-ppo-new-reward-pid-vae-sac-pid into this branch, with latest server and env.. tested ppo and sac with ppo inputs. 

Need to test: 
ppo with vae,
vae training

### CODE REVIEW

File: environment/carla_9_4/server.py [hunk 1]
```diff
@@ -12,27 +11,31 @@ def __init__(self, config=None):
...
+            os.environ["SDL_VIDEODRIVER"] = "offscreen"
         
         launch_command = [
-                self.server_binary, self.config['city_name'], "-windowed",
-                "-ResX={}".format(self.render_res_x), "-ResY={}".format(self.render_res_y),
-                "-carla-server", "-benchmark", "-fps={}".format(self.server_fps),
-                "-carla-world-port={}".format(self.server_port), "SDL_VIDEODRIVER=offscreen",
-                "SDL_HINT_CUDA_DEVICE=0"
-            ]
+                self.server_binary, "-carla-rpc-port={}".format(self.server_port)
```
Comment (agarwaltanmay) [inline environment/carla_9_4/server.py]: Make sure, this launch command is correct for benchmarking and for fps setting?

File: agents/tf/train_measurements_sac_run.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+    print("Training begins")
+    IMAGES_PATH = ALTA_LOGS+'images/'
+    VIDEO_PATH = ALTA_LOGS+'videos/'
+    vis_wrapper = vis_module.vis(IMAGES_PATH, VIDEO_PATH, FRAME_SKIP)
+    
+    # config = ConfigManager(algo="PPO")
+    logger = tf_log.Logger(TB_LOGS_DIR)
+    
+    RETRIES_ON_ERROR = 5
```
Comment (agarwaltanmay) [inline agents/tf/train_measurements_sac_run.py]: Why do we need this in the train script? Can we remove it and have it in env or server code?

File: agents/tf/train_measurements_sac_run.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            else:
+                print("specify either 1_layer or 2_layer as layers input")
+                env.close()
+                print("exiting")
+                return
+
+            model = MY_SAC(policy=policy, env=dummy_env, learning_rate=args.lr,buffer_size=args.buffer_size,batch_size=512,learning_starts=5000,
+                tensorboard_log=TB_LOGS_DIR, full_tensorboard_log=False, verbose=1)
+            
+            model.learn(env, args.timesteps, 0, tb_log_name="SAC", save_file=SAVE_PATH, reset_num_timesteps=True)
```
Comment (agarwaltanmay) [inline agents/tf/train_measurements_sac_run.py]: Can you also add resume logic during training as i did it in PPO?

File: agents/tf/ae/plot_cm.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  plt.close(figure)
+  buf.seek(0)
+  # Convert PNG buffer to TF image
+  buf_value = buf.getvalue()
+  image = tf.image.decode_png(buf_value, channels=4)
+  # Add the batch dimension
+  image = tf.expand_dims(image, 0)
+  return buf_value, image
+
+# def image_grid():
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/ae/plot_cm.py]: Ok, I will do it after merging AE branch.. don't want it to cause conflict on merging

File: agents/tf/my_sac.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    :param reuse: (bool) If the policy is reusable or not
+    :param _kwargs: (dict) Extra keyword arguments for the nature CNN feature extraction
+    """
+
+    def __init__(self, sess, ob_space, ac_space, n_env=1, n_steps=1, n_batch=None, reuse=False, **_kwargs):
+        super(My_MlpPolicy_1layer, self).__init__(sess, ob_space, ac_space, n_env, n_steps, n_batch, reuse,
+                                        layers=[64],
+                                        feature_extraction="mlp", **_kwargs)
+
+class My_MlpPolicy_2layer(FeedForwardPolicy):
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/my_sac.py]: sure.. created sac_models.py

File: agents/tf/my_sac.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        f.write("Best model: {}\n".format(best_file_name))
+        f.write("Best model appears at index: {}\n".format(ind))
+        f.write("No of successes in best model: {}\n".format(total_successes[ind]))
+        f.write("Max no of successes: {}\n".format(m))
+        f.write("Rewards at intermediate training: {}\n".format(total_rewards))
+        f.write("Total success episodes: {}\n".format(total_successes))
+
+    return best_model
+
+class MY_SAC(SAC):
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/my_sac.py]: As discussed, my_sac.py is the file similar to your ppo.py. Essentially this is extension on the base sac class as my_sac.

File: agents/tf/train_measurements_ppo.py [hunk 7]
```diff
@@ -54,7 +54,7 @@ def test(model, env):
     success_episodes = 0
     results = {}
     total_reward = 0
-    for ind in range(25):
+    for ind in range(5):
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/train_measurements_ppo.py]: done

File: agents/tf/train_measurements_ppo_run.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+#         f.write("Best model: {}\n".format(path))
+#         f.write("Best model appears at index: {}\n".format(ind))
+#         f.write("No of successes in best model: {}\n".format(total_successes[ind]))
+#         f.write("Max no of successes: {}\n".format(m))
+#         f.write("Rewards at intermediate training: {}\n".format(total_rewards))
+#         f.write("Total success episodes: {}\n".format(total_successes))
+        
+#     return best_model
+
+def run_ppo(args, prefix, config):
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/train_measurements_ppo_run.py]: done

File: agents/tf/train_measurements_sac_run.py [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        else:
+            results[ind] = 0
+    print("Results of train scenarios")
+    print(results)
+    print("Step: {0} Total Success Episodes: {1}".format(model_step, success_episodes))
+    env.logger.log_scalar('test/success_episodes', success_episodes, model_step)
+    env.logger.log_scalar('test/total_reward', total_reward, model_step)
+    return total_reward, success_episodes
+
+# def get_best_model(total_timesteps, save_file, env):
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/train_measurements_sac_run.py]: done

File: environment/carla_9_4/server.py [hunk 10]
```diff
@@ -12,27 +11,31 @@ def __init__(self, config=None):
         self.config = config
         self.server_port = config['server_port']
         self.server_binary = config['server_binary']
-        self.render_res_x = config['render_res_x']
-        self.render_res_y = config['render_res_y']
-        self.server_fps = config['server_fps']
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply environment/carla_9_4/server.py]: Nope.. not needed for new server..

File: environment/carla_9_4/server.py (hunk 1 above)
Comment (hitesh11) [reply environment/carla_9_4/server.py]: Yes, now we set fps and benchmark equivalent commands from env file itself.

File: agents/tf/train_measurements_sac_run.py (hunk 3 above)
Comment (hitesh11) [reply agents/tf/train_measurements_sac_run.py]: Yup, added a TODO.. will do later :)
Added an issue: https://github.com/auton-carla/alta/issues/37

File: agents/tf/train_measurements_sac_run.py (hunk 2 above)
Comment (hitesh11) [reply agents/tf/train_measurements_sac_run.py]: As discussed, keeping it like this.

File: agents/tf/train_measurements_sac.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    os.makedirs(TF_MODELS)
+
+IMAGES_PATH = ALTA_LOGS+prefix+'images/'
+VIDEO_PATH = ALTA_LOGS+prefix+'videos/'
+FRAME_SKIP = 4
+SAVE_PATH = ALTA_LOGS + prefix + 'sac_weights'
+TB_LOGS_DIR = ALTA_LOGS+prefix+str(datetime.now())
+
+def train():
+    model = SAC(policy=MlpPolicy, env=env, learning_rate=3e-4,buffer_size=2000000,batch_size=256,
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/train_measurements_sac.py]: yup, sorry, I am not using this file anymore.. will remove this in next commit.

File: agents/tf/vis_module.py [hunk 12]
```diff
@@ -22,7 +22,7 @@ def save_image(self, image, step_number):
     def generate_video(self, episode_number):
         vid_path = os.path.join(self.video_path, "log_{}.mp4".format(episode_number))
         im_path = os.path.join(self.images_path, "img%04d.png")
-        gen_vid_command = ["ffmpeg", "-y", "-i", im_path ,"-c:v", "libx264", "-r", "30", "-pix_fmt", "yuv420p",
+        gen_vid_command = ["ffmpeg", "-y", "-i", im_path ,"-c:v", "libx264", "-r", "25", "-pix_fmt", "yuv420p",
```
[the remark this answers is not in the document]
Comment (hitesh11) [reply agents/tf/vis_module.py]: Ok reverting it back to 30. we will review later while making videos..
